Random Batch Sampler Speedup

GalAvineri opened the pull request:
# Motivation
`Sampler` outputs indices using a generator, forcing `BatchSampler` to iterate over the indices one-by-one before grouping them into batches.
If  `Sampler` constructs the whole sequence of indices before yielding it, batching could be done more efficiently over the sequence than by iterating over a generator.

This occurs for example in the widely used `RandomSampler`. 
This PR replaces iteration with slicing by merging `RandomSampler` and `BatchSampler` into `RandomBatchSampler`.

Builds upon https://github.com/pytorch/pytorch/pull/137423
Benchmarking code is based on https://github.com/pytorch/pytorch/issues/76950

```
 batch_size  drop_last  replacement avg and std original   avg and std new  speedup
          4       True         True    0.0085 +- 3.7e-04 0.0032 +- 2.3e-04  166.63%
          4       True        False    0.0126 +- 3.2e-04 0.0041 +- 1.1e-03  210.52%
          4      False         True    0.0100 +- 2.3e-04 0.0031 +- 3.9e-05  222.77%
          4      False        False    0.0068 +- 5.0e-04 0.0037 +- 9.6e-05   85.69%
          8       True         True    0.0083 +- 1.2e-04 0.0016 +- 1.9e-05  403.48%
          8       True        False    0.0054 +- 1.2e-04 0.0022 +- 7.5e-05  147.67%
          8      False         True    0.0090 +- 9.0e-05 0.0016 +- 3.5e-05  452.94%
          8      False        False    0.0060 +- 1.2e-04 0.0022 +- 7.9e-05  172.32%
         64       True         True    0.0079 +- 1.0e-04 0.0003 +- 1.9e-05 2257.91%
         64       True        False    0.0050 +- 1.1e-04 0.0009 +- 2.0e-05  457.21%
         64      False         True    0.0082 +- 5.7e-05 0.0003 +- 1.7e-05 2418.74%
         64      False        False    0.0052 +- 8.8e-05 0.0009 +- 2.1e-05  475.84%
        256       True         True    0.0078 +- 9.2e-05 0.0002 +- 1.6e-05 3696.33%
        256       True        False    0.0052 +- 4.9e-05 0.0008 +- 2.1e-05  555.58%
        256      False         True    0.0084 +- 5.4e-05 0.0002 +- 1.1e-05 3676.29%
        256      False        False    0.0056 +- 1.2e-03 0.0008 +- 2.9e-05  601.11%
       1024       True         True    0.0082 +- 6.0e-05 0.0002 +- 1.6e-05 4226.53%
       1024       True        False    0.0052 +- 4.9e-05 0.0008 +- 1.8e-05  589.77%
       1024      False         True    0.0083 +- 7.4e-05 0.0002 +- 1.6e-05 4216.05%
       1024      False        False    0.0053 +- 7.3e-05 0.0008 +- 1.8e-05  598.53%
       4096       True         True    0.0080 +- 1.0e-04 0.0002 +- 1.9e-05 4200.74%
       4096       True        False    0.0053 +- 8.2e-05 0.0007 +- 1.5e-05  608.29%
       4096      False         True    0.0081 +- 1.3e-04 0.0002 +- 1.4e-05 4398.71%
       4096      False        False    0.0052 +- 6.9e-05 0.0007 +- 1.2e-05  604.97%
       8192       True         True    0.0079 +- 7.3e-05 0.0002 +- 1.5e-05 4324.38%
       8192       True        False    0.0053 +- 8.5e-05 0.0007 +- 1.9e-05  613.55%
       8192      False         True    0.0080 +- 5.8e-05 0.0002 +- 1.3e-05 4545.14%
       8192      False        False    0.0053 +- 1.0e-04 0.0007 +- 1.2e-05  613.44%
      16384       True         True    0.0081 +- 1.6e-04 0.0002 +- 1.1e-05 4527.95%
      16384       True        False    0.0052 +- 1.1e-04 0.0007 +- 2.2e-05  606.50%
      16384      False         True    0.0080 +- 6.5e-05 0.0002 +- 1.2e-05 4462.77%
      16384      False        False    0.0052 +- 4.0e-05 0.0007 +- 1.7e-05  604.40%
```

In order to support the `replacement` argument I used numpy's `choice` since I couldn't find an efficient alternative in pytorch.
Therefore I also used a numpy.random.Generator in the `generator` argument.

If it is required to not use numpy, I could look into finding an efficient torch alternative for `choice`.

GalAvineri committed:
Added RandomBatchSampler

GalAvineri commented:
This is the speed testing code:

```python
from typing import Sized, Iterator
import timeit
import numpy as np
import pandas as pd
from torch.utils.data import Sampler, RandomSampler, BatchSampler

class RandomBatchSampler(Sampler[list[int]]):
    def __init__(
            self,
            data_source: Sized,
            replacement: bool = False,
            generator : np.random.Generator = None,
            batch_size: int = 32,
            drop_last: bool = False,
    ) -> None:
        super().__init__()
        self.data_source = data_source
        self.replacement = replacement
        self.generator = generator
        self.batch_size = batch_size
        self.drop_last = drop_last and len(data_source) % self.batch_size > 0

        if not isinstance(self.replacement, bool):
            raise TypeError(
                f"replacement should be a boolean value, but got replacement={self.replacement}"
            )

        self.n_batches = len(data_source) // batch_size

    def initialize_indices(self) -> np.ndarray[int]:
        indices = np.arange(len(self.data_source))
        generator = self.generator if self.generator is not None else np.random.default_rng()
        indices = indices if not self.replacement else generator.choice(indices, len(indices), replace=True)
        return indices

    def __iter__(self) -> Iterator[list[int]]:
        indices = self.initialize_indices()
        indices_batches = [indices[i:i + self.batch_size] for i in range(0, len(indices), self.batch_size)]
        if self.drop_last:
            indices_batches.pop()
        yield from indices_batches

    def __len__(self):
        return self.n_batches

def _iter_on_origin_sampler(batch_size, drop_last, replacement):
    for _ in BatchSampler(RandomSampler(range(DATA_SIZE), replacement=replacement), batch_size=batch_size, drop_last=drop_last):
        pass


def _iter_on_new_sampler(batch_size, drop_last, replacement):
    for _ in RandomBatchSampler((range(DATA_SIZE)), batch_size=batch_size, drop_last=drop_last, replacement=replacement):
        pass


if __name__ == '__main__':
    DATA_SIZE = 1_000_000
    AVG_TIMES = 10

    data = np.zeros(DATA_SIZE)

    results = []
    for batch_size in [4, 8, 64, 640, 6400]:
        for drop_last in [True, False]:
            for replacement in [True, False]:
                timer = timeit.Timer(lambda: _iter_on_origin_sampler(batch_size, drop_last, replacement))
                times_original = timer.repeat(AVG_TIMES, 1)
                original_avg = np.mean(times_original)
                original_std = np.std(times_original)
                desc_original = f"{original_avg:.4f} +- {original_std:.1e}"

                timer = timeit.Timer(lambda: _iter_on_new_sampler(batch_size, drop_last, replacement))
                times_new = timer.repeat(AVG_TIMES, 1)
                new_avg = np.mean(times_new)
                new_std = np.std(times_new)
                desc_new = f"{new_avg:.4f} +- {new_std:.1e}"

                speedup_percent = "%.2f" % ((1 / new_avg - 1 / original_avg) * original_avg * 100) + "%"

                current_row = [batch_size, drop_last, replacement,
                               desc_original,
                               desc_new,
                               speedup_percent]

                results.append(current_row)


    columns = ["batch_size", "drop_last", "replacement", "avg and std original", "avg and std new", "speedup"]
    results = pd.DataFrame(results, columns=columns)

    pd.set_option('display.max_columns', None)
    pd.set_option('display.width', 1000)
    print(results.to_string(index=False))
```

GalAvineri committed:
Added shuffle in case replacement=False

vadimkantorov commented on torch/utils/data/sampler.py:
```diff
@@ -346,3 +347,45 @@ def __len__(self) -> int:
+            self,
+            data_source: Sized,
+            replacement: bool = False,
+            generator : np.random.Generator = None,
```
for generator, it is usually, torch.Generator, no? any reason for NumPy generator here?

GalAvineri committed:
Fixed type hinting

GalAvineri committed:
Accelerated in case replacement=False by avoiding unnecessary np.arange.

divyanshk commented:
Thanks for waiting on me @GalAvineri 

1) Do the speedup numbers in the PR compare `RandomBatchSampler(..)` vs ` BatchSampler(RandomSampler(..)`  ?
2) Can we add unit tests for  `RandomBatchSampler` ? 
3) Regarding supporting the replacement argument, it will be better if we can use a torch alternative instead of importing numpy, unless there are performance implications.

Thanks.

divyanshk commented on torch/utils/data/sampler.py:
```diff
@@ -346,3 +350,53 @@ def __len__(self) -> int:
+        self,
+        data_source: Sized,
+        replacement: bool = False,
+        generator: Optional[np.random.Generator] = None,
```
This probably shouldn't be optional.

GalAvineri replied:
I'll change the arguments to be exactly as in `RandomSampler` and `BatchSampler`.

divyanshk commented on torch/utils/data/sampler.py:
```diff
@@ -346,3 +350,53 @@ def __len__(self) -> int:
+
+    def sample_indices(self) -> NDArray[np.int_]:
+        generator = (
+            self.generator if self.generator is not None else np.random.default_rng()
```
In case `self.generator` is `None`, we can use `torch.Generator()`?

GalAvineri replied:
I'll remove all `numpy` from the implementation

GalAvineri committed:
Switched to torch-based implementation

GalAvineri committed:
Added unit test

GalAvineri commented:
Since https://github.com/pytorch/pytorch/pull/149441 got a bit complicated, I thought to proceed with this simpler PR.

I removed all `numpy` based implementation and replaced it with `torch` based implementation.
I also added unit tests for `RandomBatchSampler`.

@divyanshk Please let me know if there is anything else you would like :)

GalAvineri commented:
These are the speedups of the `torch` implementation:
```
                                        original(avg)  original(std)  new(avg)     new(std)    speedup     
batch_size   drop_last    replacement                                                                      
4            True         True          0.084708       0.005989       0.201306     0.009317      -57.92%   
             False        True          0.103408       0.001701       0.194938     0.002805      -46.95%   
             True         False         0.067643       0.000858       0.196751     0.002000      -65.62%   
             False        False         0.081394       0.001218       0.197983     0.002455      -58.89%   
8            True         True          0.074029       0.000288       0.095129     0.000876      -22.18%   
             False        True          0.086834       0.000859       0.094614     0.000741       -8.22%   
             True         False         0.063882       0.000758       0.100378     0.000891      -36.36%   
             False        False         0.072893       0.000395       0.100430     0.000688      -27.42%   
16           True         True          0.073173       0.000856       0.048848     0.000635       49.80%   
             False        True          0.080880       0.001238       0.048868     0.000548       65.51%   
             True         False         0.062894       0.000279       0.053685     0.000363       17.15%   
             False        False         0.067001       0.000405       0.054105     0.000586       23.84%   
32           True         True          0.071242       0.000829       0.025369     0.000587      180.82%   
             False        True          0.076153       0.000599       0.025316     0.000101      200.81%   
             True         False         0.062579       0.000660       0.030557     0.000439      104.79%   
             False        False         0.064653       0.000666       0.030807     0.000274      109.87%   
64           True         True          0.071466       0.000908       0.013869     0.000186      415.29%   
             False        True          0.075118       0.000439       0.013746     0.000212      446.46%   
             True         False         0.062312       0.000560       0.018989     0.000272      228.15%   
             False        False         0.064354       0.000397       0.019222     0.000146      234.80%   
128          True         True          0.071985       0.000819       0.008038     0.000165      795.60%   
             False        True          0.077950       0.000581       0.008013     0.000082      872.85%   
             True         False         0.061785       0.000406       0.013330     0.000217      363.50%   
             False        False         0.064397       0.000404       0.013458     0.000373      378.49%   
256          True         True          0.071140       0.000731       0.005096     0.000032     1295.87%   
             False        True          0.075263       0.000547       0.005086     0.000075     1379.79%   
             True         False         0.061171       0.000542       0.010761     0.000184      468.45%   
             False        False         0.064067       0.000592       0.010366     0.000341      518.02%   
512          True         True          0.071423       0.000652       0.003703     0.000017     1829.01%   
             False        True          0.074103       0.000849       0.003693     0.000055     1906.48%   
             True         False         0.061798       0.000553       0.007795     0.000630      692.79%   
             False        False         0.064311       0.000622       0.009438     0.000229      581.44%   
1024         True         True          0.072482       0.000513       0.003059     0.000043     2269.69%   
             False        True          0.073626       0.000880       0.003001     0.000026     2353.78%   
             True         False         0.061459       0.000624       0.008693     0.000139      606.97%   
             False        False         0.063144       0.000949       0.007042     0.000643      796.64%   
2048         True         True          0.071304       0.000524       0.002659     0.000019     2581.48%   
             False        True          0.073429       0.000493       0.002665     0.000044     2654.83%   
             True         False         0.061611       0.000485       0.007883     0.000247      681.60%   
             False        False         0.063150       0.000673       0.008040     0.000166      685.46%   
4096         True         True          0.071233       0.001192       0.002496     0.000040     2753.64%   
             False        True          0.073361       0.000695       0.002487     0.000048     2849.77%   
             True         False         0.061547       0.000515       0.008020     0.000115      667.43%   
             False        False         0.063076       0.000214       0.007653     0.000197      724.16%   
8192         True         True          0.071263       0.000690       0.002400     0.000015     2869.05%   
             False        True          0.072696       0.000742       0.002399     0.000031     2930.14%   
             True         False         0.062277       0.000521       0.007678     0.000103      711.13%   
             False        False         0.063158       0.000490       0.008070     0.000143      682.65%   
16384        True         True          0.073517       0.001171       0.002355     0.000020     3021.77%   
             False        True          0.074202       0.000612       0.002354     0.000016     3051.98%   
             True         False         0.063429       0.000339       0.007607     0.000210      733.87%   
             False        False         0.063530       0.000307       0.007732     0.000145      721.67%
```

GalAvineri committed:
Accelerated for small batch sizes

GalAvineri commented:
These are the speedups after removing usage of `numpy`:
```
                                        original(avg)  original(std)  new(avg)     new(std)    speedup     
batch_size   drop_last    replacement                                                                      
4            True         True          0.077361       0.001752       0.046196     0.001111       67.46%   
             False        True          0.096277       0.001955       0.042328     0.000543      127.45%   
             True         False         0.070717       0.001534       0.051294     0.001127       37.87%   
             False        False         0.086170       0.001569       0.046817     0.000441       84.06%   
8            True         True          0.074448       0.000945       0.036348     0.000260      104.82%   
             False        True          0.084396       0.000754       0.032137     0.000481      162.61%   
             True         False         0.067243       0.000538       0.040406     0.000324       66.42%   
             False        False         0.076455       0.000561       0.036644     0.000401      108.64%   
16           True         True          0.072844       0.001015       0.051147     0.001097       42.42%   
             False        True          0.077790       0.001318       0.049599     0.000539       56.84%   
             True         False         0.066606       0.000690       0.055055     0.001180       20.98%   
             False        False         0.071119       0.001579       0.053538     0.001417       32.84%   
32           True         True          0.071427       0.000895       0.025986     0.000430      174.86%   
             False        True          0.075948       0.001242       0.026491     0.000308      186.69%   
             True         False         0.066858       0.000910       0.031280     0.001601      113.74%   
             False        False         0.069822       0.000942       0.031740     0.002089      119.98%   
64           True         True          0.072361       0.001379       0.014195     0.000402      409.78%   
             False        True          0.075627       0.002324       0.014001     0.000331      440.14%   
             True         False         0.065734       0.001027       0.019085     0.000318      244.43%   
             False        False         0.068074       0.000818       0.019414     0.000291      250.65%   
128          True         True          0.070799       0.001911       0.008154     0.000037      768.29%   
             False        True          0.073799       0.000367       0.008164     0.000058      803.96%   
             True         False         0.064791       0.000556       0.011636     0.000836      456.81%   
             False        False         0.068400       0.000777       0.013891     0.000113      392.40%   
256          True         True          0.070954       0.000564       0.005279     0.000042     1244.17%   
             False        True          0.074482       0.000988       0.005361     0.000058     1289.45%   
             True         False         0.065680       0.000660       0.009615     0.000362      583.13%   
             False        False         0.067947       0.001012       0.008243     0.001056      724.29%   
512          True         True          0.072583       0.000923       0.003893     0.000086     1764.50%   
             False        True          0.073721       0.002031       0.003776     0.000050     1852.28%   
             True         False         0.065812       0.000692       0.008533     0.000334      671.26%   
             False        False         0.066747       0.001561       0.008604     0.000311      675.73%   
1024         True         True          0.072452       0.000382       0.003198     0.000045     2165.72%   
             False        True          0.072604       0.000823       0.003105     0.000031     2238.26%   
             True         False         0.065036       0.001010       0.007973     0.000299      715.67%   
             False        False         0.067535       0.000658       0.005608     0.000962     1104.21%   
2048         True         True          0.074144       0.001107       0.002782     0.000059     2564.97%   
             False        True          0.072833       0.001297       0.002799     0.000038     2501.97%   
             True         False         0.065317       0.000707       0.007677     0.000222      750.79%   
             False        False         0.067148       0.000610       0.008302     0.000137      708.77%   
4096         True         True          0.071696       0.000822       0.002542     0.000035     2720.77%   
             False        True          0.072413       0.001693       0.002511     0.000029     2783.91%   
             True         False         0.066336       0.000471       0.006060     0.000799      994.62%   
             False        False         0.066454       0.000905       0.007068     0.000326      840.27%   
8192         True         True          0.073054       0.001202       0.002504     0.000051     2817.67%   
             False        True          0.073117       0.001131       0.002403     0.000041     2942.23%   
             True         False         0.064256       0.001185       0.005114     0.000776     1156.54%   
             False        False         0.062578       0.006234       0.005196     0.000408     1104.37%   
16384        True         True          0.075046       0.000456       0.002383     0.000045     3049.44%   
             False        True          0.072383       0.000784       0.002357     0.000024     2970.89%   
             True         False         0.064347       0.000682       0.005994     0.000627      973.55%   
             False        False         0.064587       0.001146       0.007257     0.000292      789.99%   
```

GalAvineri committed:
Remove unnecessary comment

divyanshk commented on test/test_dataloader.py:
```diff
@@ -3615,6 +3615,32 @@ def test_out_of_order_iterable_ds(self):
+        data = range(10)
+        batch_size = 3
+
+        for replacement in [False, True]:
+            for drop_last in [False, True]:
```
Can we use `@parametrize` decorator to handle this ?

GalAvineri replied:
Of course :)

divyanshk commented on torch/utils/data/sampler.py:
```diff
@@ -191,10 +187,18 @@ def __iter__(self) -> Iterator[int]:
     def __len__(self) -> int:
         return self.num_samples
 
-
```
I think the linter would require 2 lines before `class` definition.

GalAvineri replied:
My Bad. I didn't run the linter yet. Thank you for catching that!

divyanshk commented on torch/utils/data/sampler.py:
```diff
@@ -210,7 +214,7 @@ def __init__(self, indices: Sequence[int], generator=None) -> None:
         self.generator = generator
 
     def __iter__(self) -> Iterator[int]:
-        for i in torch.randperm(len(self.indices), generator=self.generator):
```
This change will go away with a rebase on latest branch.

divyanshk commented on torch/utils/data/sampler.py:
```diff
@@ -346,3 +350,50 @@ def __len__(self) -> int:
+        self,
+        data_source: Sized,
+        replacement: bool = False,
+        batch_size: int = 32,
+        drop_last: bool = False,
```
Let's not have defaults for batch_size and drop_last.

GalAvineri replied:
Done :)

GalAvineri commented on torch/utils/data/sampler.py:
```diff
@@ -346,3 +350,50 @@ def __len__(self) -> int:
+        indices = self.sample_indices()
+
+        # Slicing is faster on list when batch size is small
+        # if self.batch_size < 16:
```
The if should not be commented. This is a typo.

divyanshk replied:
Perfect timing, was literally about to type that, thanks

GalAvineri commented on torch/utils/data/sampler.py:
```diff
@@ -191,10 +187,18 @@ def __iter__(self) -> Iterator[int]:
 
+    def init_generator(self):
+        if self.generator is None:
+            seed = int(torch.empty((), dtype=torch.int64).random_().item())
```
I originally just refactored this code an extracted it as a method.
If you'd like, I'll apply the change you ask :)

vadimkantorov replied:
just a suggestion. it's nice that `manual_seed` also returns a `Generator` :)

GalAvineri replied:
I agree with the suggestion :) Thank you!

GalAvineri committed:
Exported parameters to @parametrize

GalAvineri committed:
Added fixes

GalAvineri committed:
Merge branch 'pytorch:main' into faster-random-batch

GalAvineri committed:
Merge branch 'main' into faster-random-batch

GalAvineri committed:
Merge remote-tracking branch 'origin/faster-random-batch' into faster-random-batch

GalAvineri commented:
Thank you for the review! Let me know your thoughts about this draft :)

GalAvineri committed:
Fixed test

GalAvineri committed:
Fixed linter

GalAvineri committed:
Fixed linter

GalAvineri commented:
Thank you for the review! Let me know your thoughts about this draft :)

GalAvineri commented on torch/utils/data/sampler.py:
```diff
@@ -346,3 +347,57 @@ def __len__(self) -> int:
+        indices = self.sample_indices()
+
+        # Slicing is faster on list when batch size is small
+        if self.batch_size < 16:
+            indices = indices.tolist()  # type: ignore
```
In case `batch_size < 16` the yielded type is `list[int]`, otherwise it is `Tensor`.
Is this an issue?

GalAvineri committed:
Merge remote-tracking branch 'origin/faster-random-batch' into faster-random-batch

# Conflicts:
#	torch/utils/data/sampler.py

GalAvineri committed:
Merge remote-tracking branch 'origin/faster-random-batch' into faster-random-batch

# Conflicts:
#	torch/utils/data/sampler.py

GalAvineri committed:
Merge remote-tracking branch 'origin/faster-random-batch' into faster-random-batch

GalAvineri committed:
Merge remote-tracking branch 'origin/faster-random-batch' into faster-random-batch

GalAvineri committed:
Merge remote-tracking branch 'origin/faster-random-batch' into faster-random-batch

GalAvineri commented on torch/utils/data/sampler.py:
```diff
@@ -346,3 +347,59 @@ def __len__(self) -> int:
             return (len(self.sampler) + self.batch_size - 1) // self.batch_size  # type: ignore[arg-type]
+
+
+class RandomBatchSampler(Sampler[Union[torch.Tensor, list[int]]]):
```
In case batch_size < 16 the yielded type is list[int], otherwise it is Tensor.
Is this an issue?

divyanshk replied:
We decide on 16 based on the benchmarking ? 

> # Slicing is faster on list when batch size is small
Do you mind mentioning how much is the difference ? I'm just thinking if we can get away with just one type - it is not intuitive how the types switch unless the user goes looking at the code.

Also, do we see significant perf drop if we just do `List[int]` like in `BatchSampler`?

GalAvineri replied:
These are the speedups comparison between outputting `Tensor` and `List[int]`
```
                                 tensor speedup list speedup
batch_size drop_last replacement                            
4          True      True               -59.74%       70.91%
           False     True               -48.60%      105.75%
           True      False              -66.06%       32.98%
           False     False              -58.86%       72.66%
8          True      True               -21.46%      128.56%
           False     True                -8.93%      162.25%
           True      False              -38.05%       73.13%
           False     False              -29.65%       92.26%
16         True      True                53.87%      156.65%
           False     True                63.77%      177.13%
           True      False               15.04%       92.64%
           False     False               19.29%      107.54%
32         True      True               184.59%      171.52%
           False     True               204.93%      187.73%
           True      False               99.53%      102.54%
           False     False              106.50%      111.35%
64         True      True               418.91%      165.22%
           False     True               442.52%      169.36%
           True      False              218.32%       97.65%
           False     False              238.59%      106.67%
128        True      True               825.41%      223.34%
           False     True               855.68%      238.24%
           True      False              354.56%      127.31%
           False     False              375.22%      136.33%
256        True      True              1229.66%      201.41%
           False     True              1330.88%      217.68%
           True      False              562.71%      119.89%
           False     False              528.40%      129.21%
512        True      True              1883.99%      210.61%
           False     True              1830.45%      219.07%
           True      False              628.83%      118.31%
           False     False              744.82%      133.60%
1024       True      True              2177.97%      205.58%
           False     True              2314.08%      214.30%
           True      False              605.68%      120.53%
           False     False             1060.29%      124.61%
2048       True      True              2551.70%      208.84%
           False     True              2664.56%      208.69%
           True      False              655.49%      121.70%
           False     False              732.72%      133.70%
4096       True      True              2781.90%      208.66%
           False     True              2954.09%      209.90%
           True      False              684.58%      121.75%
           False     False              683.95%      125.92%
8192       True      True              2882.72%      215.19%
           False     True              2808.82%      209.92%
           True      False              669.49%      124.20%
           False     False              723.95%      124.53%
16384      True      True              3021.89%      219.85%
           False     True              3062.53%      218.42%
           True      False              722.83%      130.07%
           False     False              766.00%      130.34%
```
The difference between `Tensor` and `List[int]` grows as batch_size grows. 
Also, `Tensor` speedup is negative when batch_size < 16, which is why I used the list implementation in these cases.

GalAvineri commented:
@divyanshk @vadimkantorov Looking forward for your response!

GalAvineri committed:
Merge branch 'pytorch:main' into faster-random-batch

GalAvineri committed:
Output is always list[int]

GalAvineri committed:
Validate arguments without constructors

GalAvineri committed:
Merge branch 'pytorch:main' into faster-random-batch

GalAvineri committed:
Merge branch 'pytorch:main' into faster-random-batch

divyanshk reviewed (commented):
@GalAvineri just opened it back again. also can you look at a comment i added earlier in torch/utils/data/sampler.py on extra constructors

divyanshk reviewed (commented):
@GalAvineri left some comments. almost there, lets land this soon

GalAvineri commented on torch/utils/data/sampler.py:
```diff
@@ -352,3 +362,56 @@ def __len__(self) -> int:
             return (len(self.sampler) + self.batch_size - 1) // self.batch_size  # type: ignore[arg-type]
+
+
+class RandomBatchSampler(Sampler[list[int]]):
```
Done :) I referenced the the documentation of the arguments instead of duplicating it.
If you prefer it otherwise let me know!

GalAvineri commented on torch/utils/data/sampler.py:
```diff
@@ -159,12 +169,7 @@ def num_samples(self) -> int:
-            generator.manual_seed(seed)
-        else:
-            generator = self.generator
+        generator = self.generator if self.generator is not None else init_generator()
```
Done :)

GalAvineri committed:
Added `item()` to `init_generator()`

GalAvineri committed:
Added dosctring

GalAvineri committed:
Changed use of `instantiate_parametrized_tests` as decorator
